<a href="https://colab.research.google.com/github/rizwanhaidar/ADVANCED-ENCRYPTION-STANDARD-AES--128-bits/blob/master/VDA_Assignment_3_Exercise_1_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# VDA Assignment 3
## Group Members
#### Maaz Asad (s65masad)
#### Maria Yasir (s09myasi)
#### Rizwan Haidar (s83rhaid)



Following are the imported libraries used for this assignment

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
from sklearn.feature_selection import f_classif
import numpy as np
import random
import types
import warnings
warnings.filterwarnings('ignore')
import os

# Exercise 1
## (Parallel Coordinates Plot in Plotly, 11 Points)

 Plotly is a plotting library with bindings for different programming languages. It enables the creation of interactive plots based on HTML, CSS and javascript. An introduction to the parallel coordinates plot with plotly can be found here: https://plot.ly/python/parallel-coordinates-plot/

In this exercises, we will work with a Mice Protein Expression Dataset, which contains expression levels of 77 proteins, measured in the cerebral cortex of 8 classes of mice. The classes result from two genotypes (Ts65Dn, which serves as a mouse model of human down syndrome, vs. normal controls), two treatments
(injection of the drug memantine vs. a saline solution as a control), and two experimental conditions related to context fear conditioning (context-shock, which should lead to learning, vs. shock-context, in which no learning takes place). Counting all repeated measurements, there are 1080 samples overall,
some with missing data. You can find more information on the data in the corresponding publication.


### a) Use pandas to read the file Data_Cortex_Nuclear.xls provided via eCampus. Extract subgroups t-CS-s and c-CS-s. How many mice were measured for each class? (1P)

In [2]:
df = pd.read_excel("Data_Cortex_Nuclear.xls")

In [3]:
df.head()

MouseID  DYRK1A_N   ITSN1_N    BDNF_N     NR1_N    NR2A_N    pAKT_N  \
0   309_1  0.503644  0.747193  0.430175  2.816329  5.990152  0.218830   
1   309_2  0.514617  0.689064  0.411770  2.789514  5.685038  0.211636   
2   309_3  0.509183  0.730247  0.418309  2.687201  5.622059  0.209011   
3   309_4  0.442107  0.617076  0.358626  2.466947  4.979503  0.222886   
4   309_5  0.434940  0.617430  0.358802  2.365785  4.718679  0.213106   

    pBRAF_N  pCAMKII_N   pCREB_N  ...   pCFOS_N     SYP_N  H3AcK18_N  \
0  0.177565   2.373744  0.232224  ...  0.108336  0.427099   0.114783   
1  0.172817   2.292150  0.226972  ...  0.104315  0.441581   0.111974   
2  0.175722   2.283337  0.230247  ...  0.106219  0.435777   0.111883   
3  0.176463   2.152301  0.207004  ...  0.111262  0.391691   0.130405   
4  0.173627   2.134014  0.192158  ...  0.110694  0.434154   0.118481   

     EGR1_N  H3MeK4_N    CaNA_N  Genotype  Treatment  Behavior   class  
0  0.131790  0.128186  1.675652   Control  Memantine       C/S  c-CS-m  
1  0.135103  0.131119  1.743610   Control  Memantine       C/S  c-CS-m  
2  0.133362  0.127431  1.926427   Control  Memantine       C/S  c-CS-m  
3  0.147444  0.146901  1.700563   Control  Memantine       C/S  c-CS-m  
4  0.140314  0.148380  1.839730   Control  Memantine       C/S  c-CS-m  

[5 rows x 82 columns]

In [4]:
df_t_CS_s = df[df['class'] == 't-CS-s']
df_c_CS_s = df[df['class'] == 'c-CS-s']

In [5]:
# df['class'].unique()

In [6]:
print("Shape of DF for Class t-CS-s:",df_t_CS_s.shape)
print("Shape of DF for Class c-CS-s:",df_c_CS_s.shape)

Shape of DF for Class t-CS-s: (105, 82)
Shape of DF for Class c-CS-s: (135, 82)


For t-CS-s class **105** mic were studied

For c_CS_s class **135** mice were studied

### b) Create a parallel coordinates plot with plotly from the following 5 proteins: (pPKCG N, pP70S6 N, pS6 N, pGSK3B N, ARC N). Assign different colors to the two selected classes. Annotate every axis with the correct protein name. (9P)

In [7]:
# Concatenate the two dataframes for plotting
df_combined = pd.concat([df_t_CS_s, df_c_CS_s])

# Select the proteins and the class column
proteins = ['pPKCG_N', 'pP70S6_N', 'pS6_N', 'pGSK3B_N', 'ARC_N']
df_plot = df_combined[proteins + ['class']]

# Convert class to category codes for numeric coloring
df_plot['class_code'] = df_plot['class'].astype('category').cat.codes
class_mapping = dict(enumerate(df_plot['class'].astype('category').cat.categories))

# Create the parallel coordinates plot
fig = px.parallel_coordinates(
    df_plot,
    color="class_code",
    dimensions=proteins,
    ##labels={**{protein: protein for protein in proteins}, "class_code": "class"},
    color_continuous_scale=px.colors.diverging.Tealrose,
    range_color=[df_plot['class_code'].min(), df_plot['class_code'].max()]
)

fig.update_coloraxes(colorbar=dict(
    tickvals=list(class_mapping.keys()),
    ticktext=list(class_mapping.values()),
    title="Class"
))

fig.show()


### c) Explore the data by interacting with the parallel coordinates plot. Do you find anything suspicious about the data set? (1P)

Yes. The dataset shows low variance across many protein features, significant overlap between the two classes, and potential outliers. This suggests some proteins may not contribute useful discriminatory information between the groups, and further analysis (e.g., statistical tests or feature selection) is necessary to isolate meaningful patterns.

#Exercise 2
## (Comparing RadViz and Star Coordinates, 12 Points)

In the lecture, we introduced star coordinates and RadViz as alternative radial data visualization techniques. The paper rubio-sanchez-radviz-sc-2016.pdf, which is available via eCampus, provides a closer investigation of their relationship. Please read it and answer the following questions in your
own words. Remember that we will not grant even partial credit for copy-pasted text.


### a) As we mentioned in the lecture, the star coordinates, and RadViz are two popular multidimensional data visualization techniques. Why they can also be interpreted as dimensionality reduction techniques? (1P)

#### **Answer to question (a):**

Star Coordinates and RadViz can be interpreted as **dimensionality reduction techniques** because they both **map high-dimensional data onto a low-dimensional (typically 2D) space**, making it possible to visualize complex datasets. Although their primary goal is visualization, they effectively perform a **projection from an n-dimensional space to a 2D space**, which is a core idea in dimensionality reduction.

- **Star Coordinates (SC)** use a linear mapping: each data point is projected using a weighted sum of axis vectors representing the dimensions. This is similar to other linear dimensionality reduction methods like PCA.
  
- **RadViz**, on the other hand, performs a **nonlinear mapping** by interpreting data values as spring forces pulling towards anchor points (dimensions), and placing the point where these forces balance. This adds a normalization step, turning the projection into a **nonlinear transformation**.

Both techniques simplify the representation of data, reducing it from many variables to two, while still preserving structural relationships hence, they qualify as dimensionality reduction methods.

### b) In general, the authors of this paper prefer star coordinates over RadViz. Briefly explain two reasons for their preference. (2P)

#### **Answer to question (b):**

The authors generally prefer **Star Coordinates (SC)** over **RadViz** for the following two key reasons:


**1. RadViz introduces nonlinear distortions that hinder data interpretation**  
RadViz includes a **normalization step** that turns the projection into a nonlinear mapping. This step can **distort the geometry** of the data in several ways:
- It alters distances between points, making clusters appear larger or smaller depending on their position relative to the origin.

In contrast, SC uses a **linear mapping**, preserving relative distances and shapes better, leading to **more faithful visualizations** of the original data structure.


**2. SC allows more flexibility and supports linear mappings like PCA or LDA**  
Star Coordinates let analysts choose axis vectors freely, making it possible to **reproduce known linear transformations** such as **Principal Component Analysis (PCA)**, **Linear Discriminant Analysis (LDA)**, and others. This flexibility enables:
- **Better class or cluster separation**, especially in classification tasks.

RadViz lacks this flexibility due to its **fixed normalization constraint** and nonlinear nature, which makes it **incapable of reproducing these valuable linear projections**.


So in summary, the authors favor SC because it (1) preserves geometric properties more accurately and (2) supports interpretable, flexible linear mappings that aid in a wide range of analysis tasks.

### c) Despite this, the authors mention a specific use case where they consider RadViz to be superior to star coordinates. Briefly explain what this use case is and why they prefer RadViz in this case. (2P)

#### **Answer to question (c):**

The authors consider **RadViz to be superior to Star Coordinates (SC) when  visualizing data is sparse**, i.e., when data samples contain only a few (related) attributes that are considerably larger than the rest..



### **Use case: Sparse data visualization**

**Sparse data** refers to data samples where **only one or a few attribute values are significantly larger than the rest** a common scenario in fields like **bioinformatics** , where only a few genes might be highly expressed in each sample.



### **Why RadViz is preferred in this case:**
- **RadViz’s normalization step** causes data points with **dominant attributes** to be **pulled strongly toward the corresponding anchor point**, making them stand out more clearly.
- This means **sparse data points appear near the related dimension **, and **non-sparse data cluster around the center**, allowing **better visual separation** of sparse samples from the rest.
- In contrast, in SC, **sparse and non-sparse points may overlap** more, making it harder to visually detect or distinguish sparse samples.


So, although Star Coordinates are generally more flexible and preserve structure better, **RadViz excels in highlighting samples with a few standout features**, which is especially useful for detecting **dominant patterns in sparse datasets**.

### d) Neither RadViz nor star coordinates provide a one-to-one mapping between a data point’s highdimensional location and its two-dimensional projection. Interactively modifying the anchor points or axes, respectively, can reduce the resulting ambiguities. However, the paper mentions two examples in which, unlike star coordinates, RadViz introduces ambiguities that cannot be resolved with any re-arrangement of the anchor points. Point out these two examples. Could the extended RadViz that we learned about in the lecture resolve them? (3P)

#### **Answer to (d):**

The paper highlights **two cases** where RadViz creates ambiguities that **can’t be fixed**, even by changing the anchor points:

1. **Cluster size distortion**: In RadViz, clusters closer to the origin appear larger because of the normalization step. This visual distortion doesn’t reflect the true size and **can’t be corrected** by reordering anchors.  
2. **Overlapping classes**: Some classes (e.g., in the Iris dataset) always overlap in RadViz because it maps data to a **unit simplex**. This causes different classes to fall into the same area, and **no anchor layout** can separate them.

**Extended RadViz** (like 3D RadViz or versions using extra dimensions) can **partially reduce** these issues—for example, by adding the distance to the origin—but they **don’t fully solve** the problems like **Star Coordinates**, which allow linear and flexible mappings.